File: scripts/check_mvp.py
#!/usr/bin/env python3
"""Deterministic integrity and semantic checks on the MVP artifacts.

Terminology and profiles come from the SUSHI output (fsh-generated/); the single
worked example is the generated longitudinal Bundle in input/resources/.
"""
from __future__ import annotations

import base64
import json
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
RES = ROOT / "fsh-generated" / "resources"
BUNDLE_FILE = ROOT / "input" / "resources" / "Bundle-period-tracking-longitudinal-example.json"
CYCLE = "https://cycle.fhir.me/CodeSystem/cycle"
LOINC, SCT, UCUM = "http://loinc.org", "http://snomed.info/sct", "http://unitsofmeasure.org"
OBSCAT = "http://terminology.hl7.org/CodeSystem/observation-category"
FACT_PROFILE = "https://cycle.fhir.me/StructureDefinition/period-tracking-fact"
PANEL_PROFILE = "https://cycle.fhir.me/StructureDefinition/daily-tracking-panel"
BUNDLE_PROFILE = "https://cycle.fhir.me/StructureDefinition/period-tracking-bundle"
EXPECTED_CODES = {"daily-tracking-panel", "menstrual-flow", "flow-none", "flow-spotting", "flow-light", "flow-moderate", "flow-heavy"}
EXPECTED_PROFILES = {"period-tracking-bundle", "period-tracking-fact", "daily-tracking-panel"}
FLOW_VALUES = EXPECTED_CODES - {"daily-tracking-panel", "menstrual-flow"}
VALUE_KEYS = {"valueQuantity", "valueCodeableConcept", "valueString", "valueBoolean"}


def load(path: Path) -> dict[str, Any]:
    if not path.exists():
        raise AssertionError(f"Missing resource: {path}")
    return json.loads(path.read_text(encoding="utf-8"))


def has_coding(cc: dict[str, Any], system: str, code: str) -> bool:
    return any(c.get("system") == system and c.get("code") == code for c in (cc or {}).get("coding", []))


def ref_tuple(ref: str):
    if not ref or "/" not in ref or ref.startswith(("http", "urn:", "#")):
        return None
    return tuple(ref.split("/", 1))


def main() -> int:
    messages: list[str] = []
    errors: list[str] = []
    try:
        cs = load(RES / "CodeSystem-cycle.json")
        assert {c["code"] for c in cs.get("concept", [])} == EXPECTED_CODES, "project CodeSystem codes differ"
        messages.append("Project CodeSystem contains exactly the expected seven concepts.")

        vs = load(RES / "ValueSet-menstrual-flow.json")
        inc = {c["code"] for i in vs.get("compose", {}).get("include", []) if i.get("system") == CYCLE for c in i.get("concept", [])}
        assert inc == FLOW_VALUES, "Menstrual Flow ValueSet differs"
        messages.append("Menstrual Flow ValueSet contains exactly the five ordinal result codes.")

        profiles = {load(p).get("id") for p in RES.glob("StructureDefinition-*.json")}
        assert profiles == EXPECTED_PROFILES, f"profile set differs: {sorted(profiles)}"
        messages.append("Exactly three MVP profiles were generated.")

        bundle = load(BUNDLE_FILE)
        assert BUNDLE_PROFILE in bundle.get("meta", {}).get("profile", []), "bundle missing profile"
        assert bundle.get("type") == "collection" and bundle.get("identifier") and bundle.get("timestamp")
        assert "link" not in bundle and "total" not in bundle
        entries = bundle.get("entry", [])
        full_urls = [e.get("fullUrl") for e in entries]
        assert None not in full_urls and len(full_urls) == len(set(full_urls)), "fullUrls missing or not unique"
        assert all(not ({"request", "response", "search"} & set(e)) for e in entries)

        resources = [e.get("resource", {}) for e in entries]
        keys = {(r.get("resourceType"), r.get("id")) for r in resources}
        for r in resources:  # references resolve (relative ones)
            for ref in _refs(r):
                key = ref_tuple(ref)
                if key and key not in keys:
                    errors.append(f"Unresolved reference: {ref}")

        kinds = {}
        for r in resources:
            kinds.setdefault(r.get("resourceType"), []).append(r)
        assert len(kinds.get("Patient", [])) == 1, "exactly one Patient required"
        assert kinds.get("Device") and kinds.get("Provenance") and kinds.get("Binary"), "Device/Provenance/Binary required"

        facts, panels = [], []
        by_key = {(r.get("resourceType"), r.get("id")): r for r in resources}
        for obs in kinds.get("Observation", []):
            prof = set(obs.get("meta", {}).get("profile", []))
            assert obs.get("status") == "final"
            assert any(has_coding({"coding": obs.get("category", [{}])[0].get("coding", [])}, OBSCAT, c) for c in ("survey", "vital-signs")), \
                f"{obs.get('id')} category not survey/vital-signs"
            assert obs.get("subject", {}).get("reference", "").startswith("Patient/")
            assert len(obs.get("performer", [])) == 1 and obs["performer"][0]["reference"].startswith("Patient/")
            assert obs.get("device", {}).get("reference", "").startswith("Device/")
            assert "effectiveDateTime" in obs
            if PANEL_PROFILE in prof:
                panels.append(obs)
                assert has_coding(obs.get("code", {}), CYCLE, "daily-tracking-panel")
                assert not (VALUE_KEYS & set(obs)) and "component" not in obs and "derivedFrom" not in obs
                assert obs.get("hasMember") or obs.get("note"), f"empty panel {obs.get('id')}"
            elif FACT_PROFILE in prof:
                facts.append(obs)
                assert len(VALUE_KEYS & set(obs)) == 1, f"fact {obs.get('id')} must have exactly one value"
                assert not ({"hasMember", "component", "derivedFrom", "dataAbsentReason"} & set(obs))
            else:
                raise AssertionError(f"Observation {obs.get('id')} declares no MVP profile")
        assert panels and facts, "expected panels and facts"

        for p in panels:  # panel members resolve, are facts, same patient/device/day
            for m in p.get("hasMember", []):
                k = ref_tuple(m["reference"])
                assert k in by_key, f"unresolved member {m['reference']}"
                f = by_key[k]
                assert FACT_PROFILE in f.get("meta", {}).get("profile", [])
                assert f.get("subject") == p.get("subject") and f.get("device") == p.get("device")
                assert str(f.get("effectiveDateTime"))[:10] == str(p.get("effectiveDateTime"))[:10]

        for f in facts:  # per-fact value semantics
            code = f.get("code", {})
            if has_coding(code, CYCLE, "menstrual-flow"):
                vals = {c["code"] for c in f.get("valueCodeableConcept", {}).get("coding", []) if c.get("system") == CYCLE}
                assert len(vals) == 1 and vals <= FLOW_VALUES, f"bad flow value in {f.get('id')}"
            if has_coding(code, LOINC, "72514-3"):
                q = f.get("valueQuantity", {})
                assert 0 <= q.get("value", -1) <= 10 and q.get("system") == UCUM and q.get("code") == "{score}"
            if has_coding(code, LOINC, "8310-5"):
                q = f.get("valueQuantity", {})
                assert q.get("system") == UCUM and q.get("code") in {"Cel", "[degF]"}
                assert any(has_coding({"coding": cat.get("coding", [])}, OBSCAT, "vital-signs") for cat in f.get("category", [])), \
                    "temperature must be category vital-signs"
            if has_coding(code, LOINC, "8678-5"):
                vals = {c["code"] for c in f.get("valueCodeableConcept", {}).get("coding", []) if c.get("system") == SCT}
                assert vals in ({"289894009"}, {"289895005"}), f"bad menstrual status in {f.get('id')}"
        assert any(has_coding(f.get("code", {}), LOINC, "8678-5") and has_coding(f.get("valueCodeableConcept", {}), SCT, "289895005") for f in facts), \
            "expected at least one explicit-negative (not menstruating) fact"
        assert any(p.get("note") and not p.get("hasMember") for p in panels), "expected at least one note-only panel"

        prov = kinds["Provenance"][0]
        targets = {ref_tuple(t.get("reference", "")) for t in prov.get("target", [])}
        assert {("Observation", p["id"]) for p in panels} <= targets, "Provenance must target every panel"
        assert any(e.get("role") == "source" and ref_tuple(e.get("what", {}).get("reference", "")) == ("Binary", kinds["Binary"][0]["id"]) for e in prov.get("entity", [])), \
            "Provenance must cite the Binary native archive as source"

        native = json.loads(base64.b64decode(kinds["Binary"][0]["data"], validate=True))
        assert native.get("sourceApp") == "Periodicity" and native.get("days"), "native archive must parse and name the source app"
        messages.append(f"Worked Bundle: {len(panels)} daily panels, {len(facts)} facts, explicit-negative + note-only days, Provenance, and a decodable native archive.")

        all_json = list(RES.glob("*.json"))
        for r in all_json:
            json.loads(r.read_text(encoding="utf-8"))
        messages.append(f"All {len(all_json)} generated JSON resources parse successfully.")
    except (AssertionError, KeyError, TypeError, ValueError, json.JSONDecodeError) as exc:
        errors.append(f"{type(exc).__name__}: {exc}")

    report = ROOT / "validation" / "integrity-check.txt"
    report.parent.mkdir(exist_ok=True)
    report.write_text("Period Tracking MVP integrity check\n\n" + "\n".join(f"PASS: {m}" for m in messages) +
                      ("\n" if messages else "") + "\n".join(f"FAIL: {e}" for e in errors) + "\n", encoding="utf-8")
    print(report.read_text(encoding="utf-8"), end="")
    return 1 if errors else 0


def _refs(value: Any):
    if isinstance(value, dict):
        for k, v in value.items():
            if k == "reference" and isinstance(v, str):
                yield v
            else:
                yield from _refs(v)
    elif isinstance(value, list):
        for v in value:
            yield from _refs(v)


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/check-mvp.ts
#!/usr/bin/env bun
/**
 * Deterministic integrity and semantic checks on the MVP artifacts.
 *
 * Terminology and profiles come from the SUSHI output (fsh-generated/); the
 * worked example is the generated longitudinal Bundle supplied by BUNDLE_FILE
 * or produced under dist/examples/.
 */

import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = dirname(dirname(fileURLToPath(import.meta.url)));
const RES = join(ROOT, "fsh-generated", "resources");
const BUNDLE_FILE = Bun.env.BUNDLE_FILE || join(ROOT, "dist", "examples", "Bundle-period-tracking-longitudinal-example.json");
const CYCLE = "https://cycle.fhir.me/CodeSystem/cycle";
const LOINC = "http://loinc.org";
const UCUM = "http://unitsofmeasure.org";
const OBSCAT = "http://terminology.hl7.org/CodeSystem/observation-category";
const EXAMPLE_IDS = {
  bleeding: "menstrual-bleeding-example",
  flow: "menstrual-flow-example",
  symptom: "symptom-example",
  pain: "numeric-pain-severity-example",
  bbt: "basal-body-temperature-example",
};
const EXPECTED_CODES = new Set(["menstrual-bleeding", "menstrual-flow", "symptom", "flow-none", "flow-spotting", "flow-light", "flow-moderate", "flow-heavy"]);
const EXPECTED_PROFILES = new Set(["period-tracking-bundle", "period-tracking-fact", "menstrual-bleeding", "menstrual-flow", "symptom", "numeric-pain-severity", "basal-body-temperature"]);
const FLOW_VALUES = new Set(["flow-none", "flow-spotting", "flow-light", "flow-moderate", "flow-heavy"]);
const VALUE_KEYS = new Set(["valueQuantity", "valueCodeableConcept", "valueString", "valueBoolean"]);
const RESOURCE_SORT = "http://hl7.org/fhir/tools/StructureDefinition/resource-sort";
const EXPECTED_ARTIFACT_SORT = new Map([
  ["StructureDefinition/menstrual-bleeding", 10],
  ["StructureDefinition/menstrual-flow", 20],
  ["StructureDefinition/symptom", 30],
  ["StructureDefinition/numeric-pain-severity", 40],
  ["StructureDefinition/basal-body-temperature", 50],
  ["StructureDefinition/period-tracking-bundle", 60],
  ["StructureDefinition/period-tracking-fact", 70],
  ["ValueSet/menstrual-flow", 10],
  ["ValueSet/common-tracker-symptoms", 20],
  ["ValueSet/ptmvp-fact-category", 30],
  ["CodeSystem/cycle", 10],
]);

const glob = (pattern: string, cwd: string) => Array.from(new Bun.Glob(pattern).scanSync({ cwd })).sort();
const setEq = <T>(a: Set<T>, b: Set<T>) => a.size === b.size && [...a].every((x) => b.has(x));
const assert = (condition: unknown, message: string): asserts condition => {
  if (!condition) throw new Error(message);
};

async function load(path: string) {
  if (!(await Bun.file(path).exists())) throw new Error(`Missing resource: ${path}`);
  return JSON.parse(await Bun.file(path).text());
}

function hasCoding(cc: any, system: string, code: string) {
  return (cc?.coding || []).some((c: any) => c.system === system && c.code === code);
}

function factKind(obs: any): keyof typeof EXAMPLE_IDS | null {
  if (hasCoding(obs.code, CYCLE, "menstrual-bleeding")) return "bleeding";
  if (hasCoding(obs.code, CYCLE, "menstrual-flow")) return "flow";
  if (hasCoding(obs.code, CYCLE, "symptom")) return "symptom";
  if (hasCoding(obs.code, LOINC, "72514-3")) return "pain";
  if (hasCoding(obs.code, LOINC, "8310-5")) return "bbt";
  return null;
}

function refTuple(ref?: string) {
  if (!ref || !ref.includes("/") || ref.startsWith("http") || ref.startsWith("urn:") || ref.startsWith("#")) return null;
  return ref.split("/", 2).join("/");
}

function* refs(value: any): Iterable<string> {
  if (Array.isArray(value)) {
    for (const item of value) yield* refs(item);
  } else if (value && typeof value === "object") {
    for (const [key, child] of Object.entries(value)) {
      if (key === "reference" && typeof child === "string") yield child;
      else yield* refs(child);
    }
  }
}

async function main() {
  const messages: string[] = [];
  const errors: string[] = [];
  try {
    const cs = await load(join(RES, "CodeSystem-cycle.json"));
    assert(setEq(new Set((cs.concept || []).map((c: any) => c.code)), EXPECTED_CODES), "project CodeSystem codes differ");
    messages.push("Project CodeSystem contains exactly the expected eight concepts.");

    const vs = await load(join(RES, "ValueSet-menstrual-flow.json"));
    const inc = new Set<string>();
    for (const include of vs.compose?.include || []) {
      if (include.system !== CYCLE) continue;
      for (const concept of include.concept || []) inc.add(concept.code);
    }
    assert(setEq(inc, FLOW_VALUES), "Menstrual Flow ValueSet differs");
    messages.push("Menstrual Flow ValueSet contains exactly the five ordinal result codes.");

    const profiles = new Set<string>();
    for (const file of glob("StructureDefinition-*.json", RES)) profiles.add((await load(join(RES, file))).id);
    assert(setEq(profiles, EXPECTED_PROFILES), `profile set differs: ${JSON.stringify([...profiles].sort())}`);
    messages.push("Exactly seven MVP profiles were generated.");

    const ig = await load(join(RES, "ImplementationGuide-me.fhir.period-tracking-mvp.json"));
    for (const [reference, expectedSort] of EXPECTED_ARTIFACT_SORT) {
      const resource = (ig.definition?.resource || []).find((r: any) => r.reference?.reference === reference);
      assert(resource, `ImplementationGuide missing resource ${reference}`);
      const sort = (resource.extension || []).find((e: any) => e.url === RESOURCE_SORT)?.valueInteger;
      assert(sort === expectedSort, `ImplementationGuide ${reference} sort is ${sort}, expected ${expectedSort}`);
    }
    messages.push("Artifact sort metadata keeps the core profiles and terminology first.");

    const bundle = await load(BUNDLE_FILE);
    assert(!(bundle.meta?.profile?.length), "example Bundle should not stamp meta.profile");
    assert(bundle.type === "collection", "bundle must be a collection");
    const entries = bundle.entry || [];
    const fullUrls = entries.map((e: any) => e.fullUrl);
    assert(!fullUrls.includes(undefined) && new Set(fullUrls).size === fullUrls.length, "fullUrls missing or not unique");

    const resources = entries.map((e: any) => e.resource || {});
    const keys = new Set(resources.map((r: any) => `${r.resourceType}/${r.id}`));
    for (const resource of resources) {
      for (const ref of refs(resource)) {
        const key = refTuple(ref);
        if (key && !keys.has(key)) errors.push(`Unresolved reference: ${ref}`);
      }
    }

    const kinds = new Map<string, any[]>();
    for (const resource of resources) {
      const list = kinds.get(resource.resourceType) || [];
      list.push(resource);
      kinds.set(resource.resourceType, list);
    }

    const facts: any[] = [];
    for (const obs of kinds.get("Observation") || []) {
      assert(obs.status === "final", `${obs.id} status not final`);
      assert(!(obs.meta?.profile?.length), `${obs.id} should not stamp meta.profile`);
      if (obs.category) assert(["survey", "vital-signs"].some((c) => hasCoding({ coding: obs.category?.[0]?.coding || [] }, OBSCAT, c)), `${obs.id} category not survey/vital-signs`);
      assert("effectiveDateTime" in obs, `${obs.id} missing effectiveDateTime`);
      assert(factKind(obs), `Observation ${obs.id} is not a recognized MVP fact`);
      facts.push(obs);
      assert([...VALUE_KEYS].filter((k) => k in obs).length === 1, `fact ${obs.id} must have exactly one value`);
    }
    assert(facts.length, "expected facts");

    const bleedingByDate = new Map<string, boolean>();
    const flowByDate = new Map<string, string>();
    for (const fact of facts) {
      const code = fact.code;
      const date = String(fact.effectiveDateTime || "").slice(0, 10);
      if (hasCoding(code, CYCLE, "menstrual-bleeding")) {
        assert(typeof fact.valueBoolean === "boolean", `bad bleeding value in ${fact.id}`);
        bleedingByDate.set(date, fact.valueBoolean);
      }
      if (hasCoding(code, CYCLE, "menstrual-flow")) {
        const vals = new Set<string>((fact.valueCodeableConcept?.coding || []).filter((c: any) => c.system === CYCLE).map((c: any) => c.code));
        assert(vals.size === 1 && [...vals].every((v) => FLOW_VALUES.has(v)), `bad flow value in ${fact.id}`);
        flowByDate.set(date, [...vals][0]);
      }
      if (hasCoding(code, CYCLE, "symptom")) {
        assert(fact.valueCodeableConcept?.coding?.length || fact.valueCodeableConcept?.text, `bad symptom value in ${fact.id}`);
      }
      if (hasCoding(code, LOINC, "72514-3")) {
        const q = fact.valueQuantity || {};
        assert(q.value >= 0 && q.value <= 10 && q.system === UCUM && q.code === "{score}", `${fact.id} bad pain value`);
      }
      if (hasCoding(code, LOINC, "8310-5")) {
        const q = fact.valueQuantity || {};
        assert(q.system === UCUM && ["Cel", "[degF]"].includes(q.code), `${fact.id} bad temperature value`);
        assert((fact.category || []).some((cat: any) => hasCoding({ coding: cat.coding || [] }, OBSCAT, "vital-signs")), "temperature must be category vital-signs");
      }
    }
    assert([...bleedingByDate.values()].some((v) => v === true), "expected at least one bleeding=true core fact");
    assert([...bleedingByDate.values()].some((v) => v === false), "expected at least one bleeding=false core fact");
    for (const [date, flow] of flowByDate) {
      assert(bleedingByDate.has(date), `flow without bleeding core on ${date}`);
      assert(bleedingByDate.get(date) === (flow !== "flow-none"), `flow/bleeding mismatch on ${date}: ${flow}`);
    }

    assert(!kinds.get("Binary")?.length, "worked Bundle should not include a Binary native archive in this MVP");
    messages.push(`Worked Bundle: ${facts.length} recognized facts, boolean bleeding core, and flow consistency.`);

    const exampleDir = dirname(BUNDLE_FILE);
    for (const [kind, id] of Object.entries(EXAMPLE_IDS)) {
      const ex = await load(join(exampleDir, `Observation-${id}.json`));
      assert(ex.resourceType === "Observation" && ex.id === id, `${id} standalone example has wrong resource/id`);
      assert(factKind(ex) === kind, `${id} standalone example has wrong fact shape`);
      assert(!(ex.meta?.profile?.length), `${id} should not stamp meta.profile`);
      assert(!ex.text, `${id} should not include generated narrative`);
      assert(!ex.device, `${id} standalone example should not carry an unresolved device reference`);
      if (kind === "bbt") {
        assert(ex.subject?.reference === "#example-patient", `${id} standalone body-temperature example should carry a contained patient reference`);
        assert(ex.contained?.some((r: any) => r.resourceType === "Patient" && r.id === "example-patient"), `${id} standalone body-temperature example should contain its patient`);
      } else {
        assert(!ex.subject, `${id} standalone example should not carry an unresolved subject reference`);
      }
    }
    messages.push("Standalone profile examples are generated without inline profile stamps.");

    const allJson = glob("*.json", RES);
    for (const file of allJson) JSON.parse(await Bun.file(join(RES, file)).text());
    messages.push(`All ${allJson.length} generated JSON resources parse successfully.`);
  } catch (error: any) {
    errors.push(`${error?.name || "Error"}: ${error?.message || error}`);
  }

  const report = join(ROOT, "validation", "integrity-check.txt");
  await Bun.write(report, `Period Tracking MVP integrity check\n\n${messages.map((m) => `PASS: ${m}`).join("\n")}${messages.length ? "\n" : ""}${errors.map((e) => `FAIL: ${e}`).join("\n")}\n`);
  process.stdout.write(await Bun.file(report).text());
  return errors.length ? 1 : 0;
}

process.exit(await main());
